ament_tools/build_types/bazel.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ament_tools.build_type import BuildAction
+from ament_tools.build_type import BuildType
+
+from ament_tools.context import ContextExtender
+
+from ament_tools.helper import extract_argument_group
+
+
+class BazelBuildType(BuildType):
+    build_type = 'bazel'
+    description = 'bazel project'
+
+    def prepare_arguments(self, parser):
+        parser.add_argument(
+            '--bazel-args',
+            nargs='*',
+            default=[],
+            help='Arbitrary arguments which are passed to all bazel projects. '
+            "Argument collection can be terminated with '--'.")
+
+    def argument_preprocessor(self, args):
+        args, bazel_args = extract_argument_group(args, '--bazel-args')
+        extras = {
+            'bazel_args': bazel_args,
+        }
+        return args, extras
+
+    def extend_context(self, options):
+        ce = ContextExtender()
+        ce.add('bazel_args', getattr(options, 'bazel_args', []))
+        return ce
+
+    def on_build(self, context):
+        yield BuildAction(['bazel', 'build'] + context.bazel_args + ['//...'],
+                          cwd=context.source_space)
+
+    def on_install(self, context):
+        # clalancette: Note that bazel has no real concept of an install
+        # target.  Thus, we define a de-facto one here which is a run
+        # command with a target of install.
```

Comment:
Can you clarify what this target does?

Reply:
One of the projects I want to use this with is drake.  in order to do an install with drake, you run:

```
bazel run //:install /path/to/install/target
```

The code in `on_install` is doing essentially the same thing; it calls the run command with a target of `//:install` .  Unfortunately, bazel has no concept of install, so there is no universal command or target that we can use.  We can either have a de-facto standard (i.e. we require all bazel projects that want to support ament have an install target to the run command), or we allow this to be configurable with more variables in the `<export/>` section (and allow the user to specify which command/target combination does an install).  Which of those to you prefer, or do you have another idea?

Reply:
Is there a way to check with bazel if the install target this code tries to call actually exists? If that is the case I think it would be fine (that is what the CMake build type does). If that is not possible it might be a good idea to split "plain bazel" support and "ament specific bazel" support (similar to CMake and ament cmake build types).

Reply:
It looks like there is a way to query bazel for whether a target exists, so I'll implement that concept.  Thanks.